### ISSUE TITLE
Intent to Implement: Amplitude

### ISSUE THREAD

**Issue by ryanashcraft** [ampproject/amphtml#19160]
We at [Amplitude](https://amplitude.com/) are kicking off a project to build official support for AMP. This will require (1) a new set of APIs on our side and (2) a contribution to this project in order to make Amplitude a builtin analytics vendor option.

The exact API details are still being worked through, but I don't foresee them dramatically changing from this proposal. I am mostly modeling the API after Segment, Google Analytics, and mParticle's implementations. 

Taken from our current proof-of-concept, here's an example amp-analytics Amplitude config that sends page view and click events with user and event properties:

```
<amp-analytics type="amplitude" id="amplitude">
  <script type="application/json">
  {
    "vars": {
      "apiKey": "AMPLITUDE_API_KEY"
    },
    "triggers": {
      "click": {
        "on": "click",
        "selector": "#test1",
        "request": "event",
        "vars": {
          "eventType": "click"
        },
        "extraUrlParams": {
          "eventProperties:button id": "test1"
        }
      },
      "pageView": {
        "on": "visible",
        "request": "event",
        "vars": {
          "eventType": "page view"
        },
        "extraUrlParams": {
          "eventProperties:page id": "analytics vendors",
          "userProperties:amp user": "true"
        }
      }
    }
  }
  </script>
</amp-analytics>
```

Just like Segment's implementation, we plan to have our implementation store a prefixed cookie using the `cookies` config option. In addition, we plan to leverage the `linkers` field to support cross-domain session tracking.

**Comment by aghassemi** [comment]
Thanks @ryanashcraft for creating an ITI.
/to @jeffehobbs and @zhouyx for review of ITI.

**Comment by ryanashcraft** [comment]
@aghassemi @jeffehobbs @zhouyx Hi all. Just curious on the expected timeline for a green light or feedback here. Would like to be able to set appropriate expectations with my team and our customers. Thanks!

**Comment by zhouyx** [comment]
Thank you @ryanashcraft for the ITI. Yes please! Please refer to the [documentation](https://github.com/ampproject/amphtml/blob/master/extensions/amp-analytics/integrating-analytics.md) here for more implementation instructructions. 

@calebcordry Can help you with the linker settings if there's any question. Thank you!

**Comment by calebcordry** [comment]
Happy to help with any linker questions! Looking forward to the contribution!

**Comment by ryanashcraft** [comment]
Thanks all. @calebcordry, I just noticed https://github.com/ampproject/amphtml/pull/19102 - this would be extremely helpful. Any reservations using this new `useBody` feature for Amplitude? This would make nested the nature of our API much more natural and familiar to our customers:

```
        "extraUrlParams": {
          "event_type": "page view",
          "event_properties": {
            "page id": "analytics vendors"
          },
          "user_properties": {
            "user": "bar"
          },
          "groups": {
            "company": "Google"
          }
        }
```

vs

```
        "extraUrlParams": {
          "eventProperties:page id": "analytics vendors",
          "userProperties:blah": "bar",
          "groups:company": "Google"
        }
```

**Comment by calebcordry** [comment]
@ryanashcraft I think you should go for it! (it looks like you might be the first to use this new feature, so happy to help along the way)

**Comment by ryanashcraft** [comment]
@calebcordry Awesome. I’ll play aorund with it. I did notice that variable substitution wasn’t working as I expected - it’s not doing any substitution. Should this work?

```
"extraUrlParams": {
  "event_properties": {
    "title": "${title}"
  }
}
```

**Comment by calebcordry** [comment]
@ryanashcraft We do not currently support macro resolution outside of the first level. I would imagine that as `useBody` becomes more popular this will be a more common feature request. I will think through how we might do this and discuss it with the team. cc/ @jeffjose

**Comment by ryanashcraft** [comment]
@calebcordry Hmm I think that will be a blocker to us adopting this approach. For what it’s worth, we’d just need it support two levels of depth, or one level deep within certain top-level fields (e.g. all fields within `event_properties`). If you all think it’s possible to support and can recommend an approach, I’d be happy to help!

**Comment by calebcordry** [comment]
@ryanashcraft Thanks for the offer. The code that handles this logic lives here https://github.com/ampproject/amphtml/blob/ef1c5950bd607a5d441818a74c1baa10cece94ab/extensions/amp-analytics/0.1/requests.js#L343-L365

You can see that right now we are checking that the `typeof value === 'string'` you could maybe reconstruct this so that it would keep going  recursively for objects. There are also two relevant methods `variableService.expandTemplate()` that is responsible for converting the `${}` formatted macros and looking up anywhere else they may exist in your config, and the `urlReplacements.expandStringAsync` which actually resolves many of the platform level macros. 

Let me know if you want to take a stab at it and I am happy to help along the way. Otherwise we as a team will try to prioritize it appropriately.

**Comment by jeffjose** [comment]
This is great discussion. Thank you both!

@ryanashcraft - Do you think you can take a first crack at this? We can start with limiting it to 2 levels, and look to relax it later if there's more usecases we need to support.

Jeff

**Comment by zhouyx** [comment]
I would like to point out a few things here. 

- Two levels of nested object sounds reasonable to me when the object is appended to body. But we want to make sure that the solution also works with image request or when `useBody` value is `false. There're also the case of batching and parentPostMessage that we need to take into consideration.

- If we support nested `extraUrlParams` we will also need to understand how `extraUrlParamsReplaceMap` work along with the change.

**Comment by calebcordry** [comment]
These are great points @zhouyx. I have a few ideas for these concerns, but it is involved enough I think it would benefit from a group discussion. 

@ryanashcraft we are happy to have a small group meeting to discuss some of the details and design before you get started. Eventually we may want to present it in our [weekly design review](https://github.com/ampproject/amphtml/blob/master/CONTRIBUTING.md#weekly-design-reviews). Please let us know what you think. Thanks!

**Comment by ryanashcraft** [comment]
@calebcordry I have something working for 2-level depth template expansion. How about I start a PR and continue the discussion there? Does that sound good?

Thank you all for working with me and raising these concerns. Happy to work through whatever process is required to work out the details and reach a solution. In the meantime, I do need some clarifications regarding the concerns already raised (cc @zhouyx):

> But we want to make sure that the solution also works with image request or when useBody value is `false.

Currently for the cases when useBody is false or it's an image request, objects get stringified into `[object Object]` today, right? Nested objects are essentially completely unsupported at this time. Is this a behavior that is impacted by this change, or is this change going to be blocked on changing that behavior? To me it seems like a completely separate issue.

> There're also the case of batching and parentPostMessage that we need to take into consideration.

It'd be helpful to me if you could describe in more detail how batching and parentPostMessage work with template expansion (links to lines of code to review would be really helpful), and how this change could affect those features.

> If we support nested extraUrlParams we will also need to understand how extraUrlParamsReplaceMap work along with the change.

With Amplitude's use case, we wouldn't want nested object keys to be affected by `extraUrlParamsReplaceMap`. Nested objects are really just JSON-like values, with a flexible schema defined by the customer. Changing the behavior of `extraUrlParamsReplaceMap` to replace nested keys would be a problem for us if we did decide to use this feature, as we wouldn't want these replacements to affect anything but the top-level keys.

**Comment by calebcordry** [comment]
@ryanashcraft we talked as a team and thought it might be beneficial to talk about some of these things ahead of time. Our main concern is preventing you having done any wasted work.

There are a few design decisions that need to be made here (you have touched on a some of them). I think a quick meeting with the analytics team could help jumpstart the process. Does that sound reasonable to you? In the meantime I could compile a quick doc around some background information and potential solutions we have discussed internally.

**Comment by ryanashcraft** [comment]
@calebcordry Sounds good to me. Let me know what the next steps are there.

That said, one thing I'm concerned about is the timeline. It'd be really helpful for me to know usually how long it takes to work through the process for a change like this (taking into account that I'm happy to do whatever it takes to help expedite the process). Even a rough estimate would be really helpful for me to evaluate alternatives.

**Comment by jeffjose** [comment]
@ryanashcraft We've been talking about this a bunch in our internal meetings. I think the changes you're suggesting is something we want to accommodate but at the same time make sure that we cover all the bases so that there isn't a lot of back and forth.

Will it be possible to write a one-pager and get together for a quick meeting? The purpose of the meeting is to make sure we're all on the same page on what changes are going to be made. We can also use this opportunity to answer questions related to existing codebase you might have.

In regards to timeline, once we are in agreement the PR merge to production is ~2 weeks.

**Comment by ryanashcraft** [comment]
@jeffjose Sounds good. Are you asking me to write a one-pager?

**Comment by jeffjose** [comment]
Yup! @calebcordry said he can make time tomorrow (Wednesday, Nov 28) so if there's sometime that works for you we can quickly do a video call.

You can also find us over at Slack (amphtml.slack.com). Join using this [link](http://bit.ly/amp-slack-signup)

**Comment by ryanashcraft** [comment]
@jeffjose Signed up via that link. Tomorrow should work for me. I guess I'll coordinate with you and @calebcordry on Slack for a time.

Can you give me a high level overview of what your expecting from this one-pager?
